auto-farm/src/user_rewards.rs

```diff
@@ -0,0 +1,169 @@
+use common_structs::PaymentsVec;
+use mergeable::Mergeable;
+
+use crate::address_to_id_mapper::AddressId;
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+#[derive(TypeAbi, TopEncode, TopDecode)]
+pub struct RewardsWrapper<M: ManagedTypeApi> {
+    pub opt_locked_tokens: Option<EsdtTokenPayment<M>>,
+    pub other_tokens: UniquePayments<M>,
+}
+
+impl<M: ManagedTypeApi> Default for RewardsWrapper<M> {
+    #[inline]
+    fn default() -> Self {
+        Self {
+            opt_locked_tokens: None,
+            other_tokens: UniquePayments::default(),
+        }
+    }
+}
+
+#[derive(TypeAbi, TopEncode, TopDecode, NestedEncode, NestedDecode, Clone)]
+pub struct UniquePayments<M: ManagedTypeApi> {
+    payments: PaymentsVec<M>,
+}
+
+impl<M: ManagedTypeApi> Default for UniquePayments<M> {
+    #[inline]
+    fn default() -> Self {
+        Self {
+            payments: PaymentsVec::new(),
+        }
+    }
+}
+
+impl<M: ManagedTypeApi> UniquePayments<M> {
+    #[inline]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[inline]
+    pub fn new_from_payments(payments: PaymentsVec<M>) -> Self {
+        UniquePayments { payments }
+    }
+
+    pub fn add_payment(&mut self, new_payment: EsdtTokenPayment<M>) {
+        if new_payment.amount == 0 {
+            return;
+        }
+
+        let len = self.payments.len();
+        for i in 0..len {
+            let mut current_payment = self.payments.get(i);
+            if current_payment.can_merge_with(&new_payment) {
+                current_payment.amount += new_payment.amount;
+                let _ = self.payments.set(i, &current_payment);
+
+                return;
+            }
+        }
+
+        self.payments.push(new_payment);
+    }
+
+    #[inline]
+    pub fn into_payments(self) -> PaymentsVec<M> {
+        self.payments
+    }
+}
+
+impl<M: ManagedTypeApi> Mergeable<M> for UniquePayments<M> {
+    #[inline]
+    fn can_merge_with(&self, _other: &Self) -> bool {
+        true
+    }
+
+    fn merge_with(&mut self, mut other: Self) {
+        self.error_if_not_mergeable(&other);
+
+        if self.payments.is_empty() {
+            self.payments = other.payments;
+            return;
+        }
+        if other.payments.is_empty() {
+            return;
+        }
+
+        let first_len = self.payments.len();
+        let mut second_len = other.payments.len();
+        for i in 0..first_len {
+            let mut current_payment = self.payments.get(i);
+            for j in 0..second_len {
+                let other_payment = other.payments.get(j);
+                if !current_payment.can_merge_with(&other_payment) {
+                    continue;
+                }
+
+                current_payment.amount += other_payment.amount;
+                let _ = self.payments.set(i, &current_payment);
+
+                other.payments.remove(j);
```

Comment:
Doesn't removing the element during the for loop affect the next get sequence?

Reply:
We always break after removing, so in this case, it's fine.